src/recipe/recipe.controller.spec.ts

```diff
@@ -1,18 +1,175 @@
 import { RecipeController } from './recipe.controller';
+import { PrismaService } from 'src/prisma/prisma.service';
 
-describe('RecipeController', () => {
-  let controller: RecipeController;
+describe("Recipe Controller's tests", () => {
+    let recipeController: RecipeController;
+    let recipeService: RecipeService;
+    let prismaService: PrismaService;
 
-  beforeEach(async () => {
-    const module: TestingModule = await Test.createTestingModule({
-      controllers: [RecipeController],
-    }).compile();
+    const mockRecipe = {
+        id: expect.any(Number),
+        createdAt: expect.any(Number),
+        title: expect.any(String),
+        description: expect.any(String),
+        ingredients: expect.any(String),
+        preparation: expect.any(String),
+    };
 
-    controller = module.get<RecipeController>(RecipeController);
-  });
+    class mockRecipeRequest {
+        title: string;
+        description: string;
+        ingredients: string;
+        preparation: string;
+    }
 
-  it('should be defined', () => {
-    expect(controller).toBeDefined();
-  });
+    const mockRecipeService = {
+        createRecipe: jest.fn((dto: mockRecipeRequest) => {
+            return {
+                id: 1,
+                createdAt: Date.now(),
+                ...dto,
+            };
+        }),
+        fetchRecipe: jest.fn((id: number) => {
+            return {
+                id,
+                createdAt: Date.now(),
+                title: 'Fetched recipe title',
+                description: 'Fetched recipe description',
+                ingredients: 'Fetched recipe ingredients',
+                preparation: 'Fetched recipe preparation',
+            };
+        }),
+        fetchAllRecipes: jest.fn(() => {
+            return [
+                {
+                    id: 0,
+                    createdAt: Date.now(),
+                    title: 'Fetched recipe title',
+                    description: 'Fetched recipe description',
+                    ingredients: 'Fetched recipe ingredients',
+                    preparation: 'Fetched recipe preparation',
+                },
+                {
+                    id: 1,
+                    createdAt: Date.now(),
+                    title: 'Fetched recipe title',
+                    description: 'Fetched recipe description',
+                    ingredients: 'Fetched recipe ingredients',
+                    preparation: 'Fetched recipe preparation',
+                },
+            ];
+        }),
+        updateRecipe: jest.fn((id: number, dto: mockRecipeRequest) => {
+            return {
+                id,
+                createdAt: Date.now(),
+                title: dto.title || 'latest title',
+                description: dto.description || 'latest description',
+                ingredients: dto.ingredients || 'latest ingredients',
+                preparation: dto.preparation || 'latest preparation',
+            };
+        }),
```

Comment:
as the original method is `async` - double check if you shouldn't return a `Promise` here:
```suggestion
        updateRecipe: jest.fn((id: number, dto: mockRecipeRequest) => {
            return Promise.resolve({
                id,
                createdAt: Date.now(),
                title: dto.title || 'latest title',
                description: dto.description || 'latest description',
                ingredients: dto.ingredients || 'latest ingredients',
                preparation: dto.preparation || 'latest preparation',
            });
        }),
```

you can test it by trying to call the method within the test without `await` - if it returns correctly mocked data instead of a pending response then perhaps we should do the above

Reply:
In the original method we return a Promise, so I'd advise to return a Promise in a mocked method too to be consistent. And yes, all other methods would have to be updated.

It would seem like with current approach we can mock async method with a non-async one, but wuth using jest.spyOn I guess this won't work